tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,77 @@
+//! Tendermint Websocket event listener client
+
+use crate::{
+    net,
+    rpc::Request,
+    rpc::{endpoint::subscribe, Error as RPCError},
+};
+use async_tungstenite::{tokio::connect_async, tokio::TokioAdapter, tungstenite::Message};
+use futures::prelude::*;
+use tokio::net::TcpStream;
+
+/// Event Listener over webocket. https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(
+        address: net::Address,
+    ) -> Result<EventListener,RPCError> {
+        let (host, port) = match &address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(
+                    RPCError::invalid_params(&format!("invalid RPC address: {:?}", other)).into(),
+                );
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: String) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.to_owned()).into_json(),
```

Comment:
_newbie question:_ isn't query already has ownership over data? why call `to_owned()`

Reply:
Yeah, you are right there is no need for calling `to_owned` here with that method signature (with `query: String` as param). Zaki probably had a reference based signature (e.g. `query: &str`) first and forgot to remove the (now obsolete) `to_owned`.

Reply:
Did clippy not catch that?